File: Packages/OmniaApplication/Sources/OmniaApplication/AttachmentService.swift
import Foundation
import OmniaDomain

/// Stages picker input into app-owned storage, validates model capabilities,
/// resolves bounded request payloads, and owns attachment cleanup.
public actor AttachmentService {
    let storage: any AttachmentStorageProtocol
    let limits: AttachmentLimits
    let loadFile: @Sendable (URL, Int) async throws -> AttachmentImportCandidate
    let prepare: @Sendable (AttachmentImportCandidate) throws -> PreparedAttachmentContent
    let extractText: @Sendable (MessageAttachment, Data, Int) throws -> String
    let effectiveSupport: @Sendable (
        Capability,
        ProviderModelSelection
    ) async throws -> ModelCapabilitySupport

    public init(
        storage: any AttachmentStorageProtocol,
        limits: AttachmentLimits = AttachmentLimits(),
        loadFile: @escaping @Sendable (URL, Int) async throws -> AttachmentImportCandidate,
        prepare: @escaping @Sendable (AttachmentImportCandidate) throws -> PreparedAttachmentContent,
        extractText: @escaping @Sendable (MessageAttachment, Data, Int) throws -> String,
        effectiveSupport: @escaping @Sendable (
            Capability,
            ProviderModelSelection
        ) async throws -> ModelCapabilitySupport
    ) {
        self.storage = storage
        self.limits = limits
        self.loadFile = loadFile
        self.prepare = prepare
        self.extractText = extractText
        self.effectiveSupport = effectiveSupport
    }

    public func stageFiles(
        _ urls: [URL],
        existing: [MessageAttachment]
    ) async throws -> [MessageAttachment] {
        var candidates: [AttachmentImportCandidate] = []
        candidates.reserveCapacity(urls.count)
        for url in urls {
            do {
                candidates.append(try await loadFile(url, limits.maximumFileBytes))
            } catch let error as AttachmentError {
                throw error
            } catch {
                throw AttachmentError.unreadable(fileName: Self.safeName(url.lastPathComponent))
            }
        }
        return try await stage(candidates, existing: existing)
    }

    /// Atomically stages a picker batch. If one item fails, every file created
    /// by this batch is removed and the caller's existing staged set is intact.
    public func stage(
        _ candidates: [AttachmentImportCandidate],
        existing: [MessageAttachment]
    ) async throws -> [MessageAttachment] {
        guard existing.count + candidates.count <= limits.maximumCount else {
            throw AttachmentError.tooManyFiles(limit: limits.maximumCount)
        }
        var aggregate = 0
        for attachment in existing {
            guard attachment.byteCount >= 0 else {
                throw AttachmentError.storageUnavailable
            }
            let (next, overflow) = aggregate.addingReportingOverflow(
                attachment.byteCount
            )
            guard !overflow, next <= limits.maximumAggregateBytes else {
                throw AttachmentError.aggregateTooLarge(
                    limit: limits.maximumAggregateBytes
                )
            }
            aggregate = next
        }
        var result = existing
        var newlyStored: [MessageAttachment] = []
        do {
            for candidate in candidates {
                let safeName = Self.safeName(candidate.fileName)
                guard !candidate.data.isEmpty else {
                    throw AttachmentError.empty(fileName: safeName)
                }
                guard candidate.data.count <= limits.maximumFileBytes else {
                    throw AttachmentError.fileTooLarge(
                        fileName: safeName,
                        limit: limits.maximumFileBytes
                    )
                }
                let prepared = try prepare(
                    AttachmentImportCandidate(
                        data: candidate.data,
                        fileName: safeName,
                        declaredMediaType: candidate.declaredMediaType
                    )
                )
                guard !prepared.data.isEmpty else {
                    throw AttachmentError.empty(fileName: Self.safeName(prepared.fileName))
                }
                guard prepared.data.count <= limits.maximumFileBytes else {
                    throw AttachmentError.fileTooLarge(
                        fileName: Self.safeName(prepared.fileName),
                        limit: limits.maximumFileBytes
                    )
                }
                let (nextAggregate, overflow) = aggregate.addingReportingOverflow(
                    prepared.data.count
                )
                guard !overflow,
                      nextAggregate <= limits.maximumAggregateBytes
                else {
                    throw AttachmentError.aggregateTooLarge(
                        limit: limits.maximumAggregateBytes
                    )
                }
                aggregate = nextAggregate
                if try await containsDuplicate(prepared.data, in: result) {
                    throw AttachmentError.duplicate(fileName: prepared.fileName)
                }
                let identity = AttachmentIdentity()
                let key = try await storage.store(
                    prepared.data,
                    identity: identity,
                    fileExtension: prepared.fileExtension
                )
                let attachment = MessageAttachment(
                    identity: identity,
                    fileName: prepared.fileName,
                    mediaType: prepared.mediaType,
                    kind: prepared.kind,
                    byteCount: prepared.data.count,
                    storageKey: key
                )
                result.append(attachment)
                newlyStored.append(attachment)
            }
            return result
        } catch {
            for attachment in newlyStored {
                try? await storage.remove(attachment.storageKey)
            }
            if let attachmentError = error as? AttachmentError {
                throw attachmentError
            }
            throw AttachmentError.storageUnavailable
        }
    }

    private func containsDuplicate(
        _ data: Data,
        in attachments: [MessageAttachment]
    ) async throws -> Bool {
        for attachment in attachments where attachment.byteCount == data.count {
            let stored = try await storage.data(
                for: attachment.storageKey,
                maximumByteCount: limits.maximumFileBytes
            )
            if stored == data { return true }
        }
        return false
    }

    static func capability(for kind: AttachmentKind) -> Capability {
        switch kind {
        case .image: .vision
        case .pdf, .plainText: .documentInput
        }
    }

    static func safeName(_ value: String) -> String {
        let leaf = value.replacingOccurrences(of: "\\", with: "/")
            .split(separator: "/").last.map(String.init) ?? "Attachment"
        let trimmed = leaf.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? "Attachment" : String(trimmed.prefix(160))
    }
}

File: Android/core/application/src/main/kotlin/com/omnia/application/AttachmentService.kt
package com.omnia.application

import com.omnia.domain.AttachmentError
import com.omnia.domain.AttachmentIdentity
import com.omnia.domain.AttachmentKind
import com.omnia.domain.AttachmentPayload
import com.omnia.domain.AttachmentStorageProtocol
import com.omnia.domain.Capability
import com.omnia.domain.MessageAttachment
import com.omnia.domain.ModelCapabilitySupport
import com.omnia.domain.ProviderModelSelection
import com.omnia.domain.ResolvedAttachment

/**
 * Attachment service — coordinates staging, validation, resolution,
 * and cleanup of file attachments.
 *
 * Atomicity: [stage] tracks all newly stored files and rolls back
 * (removes them from storage) if any candidate fails. The existing
 * staged set is never mutated on failure.
 */
class AttachmentService(
    private val storage: AttachmentStorageProtocol,
    private val limits: AttachmentLimits = AttachmentLimits(),
    private val effectiveSupport: suspend (Capability, ProviderModelSelection) -> ModelCapabilitySupport = { _, _ -> ModelCapabilitySupport.unknown },
) {
    /**
     * Atomically stages a batch of candidates. Validates count, per-file
     * size, aggregate size, emptiness. Returns the staged attachments.
     */
    @Throws(AttachmentError::class)
    suspend fun stage(
        candidates: List<AttachmentImportCandidate>,
        existing: List<MessageAttachment> = emptyList(),
    ): List<MessageAttachment> {
        val totalCount = existing.size + candidates.size
        if (totalCount > limits.maximumCount) {
            throw AttachmentError.TooManyFiles(limits.maximumCount)
        }

        val staged = mutableListOf<MessageAttachment>()
        val newlyStored = mutableListOf<String>()

        try {
            for (candidate in candidates) {
                if (candidate.data.isEmpty()) {
                    throw AttachmentError.Empty(candidate.fileName)
                }
                if (candidate.data.size > limits.maximumFileBytes) {
                    throw AttachmentError.FileTooLarge(candidate.fileName, limits.maximumFileBytes)
                }

                val totalBytes = existing.sumOf { it.byteCount } +
                    staged.sumOf { it.byteCount } + candidate.data.size
                if (totalBytes > limits.maximumAggregateBytes) {
                    throw AttachmentError.AggregateTooLarge(limits.maximumAggregateBytes)
                }

                val identity = AttachmentIdentity(id = generateId())
                val kind = detectKind(candidate.fileName, candidate.declaredMediaType)
                val ext = fileExtension(candidate.fileName)

                val storageKey = storage.store(candidate.data, identity, ext)
                newlyStored.add(storageKey)

                staged.add(
                    MessageAttachment(
                        identity = identity,
                        fileName = normalizeFileName(candidate.fileName),
                        mediaType = candidate.declaredMediaType ?: "application/octet-stream",
                        kind = kind,
                        byteCount = candidate.data.size,
                        storageKey = storageKey,
                    )
                )
            }
            return existing + staged
        } catch (e: Exception) {
            // Rollback: remove all newly stored files
            for (key in newlyStored) {
                try {
                    storage.remove(key)
                } catch (_: Exception) { }
            }
            throw e
        }
    }

    /**
     * Validates staged attachments against limits AND the specific
     * provider/model route's capability support.
     */
    @Throws(AttachmentError::class)
    suspend fun validate(
        attachments: List<MessageAttachment>,
        selection: ProviderModelSelection,
    ) {
        if (attachments.size > limits.maximumCount) {
            throw AttachmentError.TooManyFiles(limits.maximumCount)
        }
        val totalBytes = attachments.sumOf { it.byteCount }
        if (totalBytes > limits.maximumAggregateBytes) {
            throw AttachmentError.AggregateTooLarge(limits.maximumAggregateBytes)
        }
        for (attachment in attachments) {
            val requiredCapability = capabilityFor(attachment.kind)
            val support = effectiveSupport(requiredCapability, selection)
            when (support) {
                ModelCapabilitySupport.supported -> { /* OK */ }
                ModelCapabilitySupport.unsupported -> throw AttachmentError.CapabilityUnsupported(attachment.kind)
                ModelCapabilitySupport.unknown -> throw AttachmentError.CapabilityUnknown(attachment.kind)
            }
        }
    }

    /**
     * Resolves durable attachment references into transient payloads.
     * Reads bytes from storage, creates Image or ExtractedText payloads.
     */
    @Throws(AttachmentError::class)
    suspend fun resolve(
        attachments: List<MessageAttachment>,
    ): List<ResolvedAttachment> {
        return attachments.map { attachment ->
            val data = try {
                storage.data(attachment.storageKey, limits.maximumFileBytes)
            } catch (e: Exception) {
                throw AttachmentError.Unreadable(attachment.fileName)
            }

            val payload = when (attachment.kind) {
                AttachmentKind.image -> AttachmentPayload.Image(
                    data = data,
                    mediaType = attachment.mediaType,
                )
                AttachmentKind.pdf, AttachmentKind.plainText -> AttachmentPayload.ExtractedText(
                    text = String(data, Charsets.UTF_8),
                )
            }

            ResolvedAttachment(attachment = attachment, payload = payload)
        }
    }

    /** Removes attachment bytes from storage. */
    suspend fun remove(attachment: MessageAttachment) {
        try {
            storage.remove(attachment.storageKey)
        } catch (_: Exception) { }
    }

    /** Removes multiple attachments. */
    suspend fun remove(attachments: List<MessageAttachment>) {
        for (attachment in attachments) {
            remove(attachment)
        }
    }

    /** Finds and removes stored files not referenced by any persisted message. */
    suspend fun cleanupOrphans(referencedBy: List<MessageAttachment>): Int {
        val referencedKeys = referencedBy.map { it.storageKey }.toSet()
        val allKeys = storage.allStorageKeys()
        val orphans = allKeys - referencedKeys
        for (key in orphans) {
            try {
                storage.remove(key)
            } catch (_: Exception) { }
        }
        return orphans.size
    }

    companion object {
        fun capabilityFor(kind: AttachmentKind): Capability = when (kind) {
            AttachmentKind.image -> Capability.vision
            AttachmentKind.pdf, AttachmentKind.plainText -> Capability.documentInput
        }

        fun normalizeFileName(fileName: String): String {
            val name = fileName.substringAfterLast('/').substringAfterLast('\\')
            return name.trim().take(160)
        }

        fun detectKind(fileName: String, mediaType: String?): AttachmentKind {
            val ext = fileExtension(fileName).lowercase()
            val mime = mediaType?.lowercase() ?: ""
            return when {
                ext in IMAGE_EXTENSIONS || mime.startsWith("image/") -> AttachmentKind.image
                ext == "pdf" || mime == "application/pdf" -> AttachmentKind.pdf
                ext in TEXT_EXTENSIONS || mime.startsWith("text/") -> AttachmentKind.plainText
                else -> AttachmentKind.plainText // Default to text for extraction
            }
        }

        fun fileExtension(fileName: String): String {
            val name = fileName.substringAfterLast('/').substringAfterLast('\\')
            val dotIndex = name.lastIndexOf('.')
            return if (dotIndex > 0 && dotIndex < name.length - 1) {
                name.substring(dotIndex + 1)
            } else {
                ""
            }
        }

        private val IMAGE_EXTENSIONS = setOf("jpg", "jpeg", "png", "gif", "webp", "heic", "heif")
        private val TEXT_EXTENSIONS = setOf("txt", "md", "csv", "json", "xml", "yaml", "yml", "toml")
    }

    private fun generateId(): String = java.util.UUID.randomUUID().toString()
}
